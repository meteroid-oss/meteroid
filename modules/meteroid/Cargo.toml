[package]
name = "meteroid"
version = "0.1.0"
rust-version.workspace = true
edition.workspace = true
license.workspace = true

[dependencies]
error-stack.workspace = true
thiserror.workspace = true
blake3.workspace = true
cached = { workspace = true, features = ["async", "tokio"] }
chrono = { workspace = true, features = ["clock"] }
common-build-info = { workspace = true }
common-config = { workspace = true }
common-logging = { workspace = true }
common-grpc = { workspace = true, features = ["server"] }
common-grpc-error-as-tonic-macros = { workspace = true }
common-grpc-error-as-tonic-macros-impl = { workspace = true }
meteroid-grpc = { path = "crates/meteroid-grpc", features = ["server"] }
meteroid-repository = { path = "crates/meteroid-repository", features = [] }
cornucopia_async = { workspace = true, features = ["with-serde_json-1", "deadpool"] }
deadpool-postgres.workspace = true
dotenvy.workspace = true
envconfig.workspace = true
futures.workspace = true
log.workspace = true
nanoid.workspace = true
prost-types.workspace = true
prost.workspace = true
reqwest = { workspace = true, features = ["default"] }
reqwest-middleware = { workspace = true }
reqwest-retry = { workspace = true }
segment.workspace = true
serde_json.workspace = true
serde.workspace = true
time.workspace = true
tokio = { workspace = true, features = ["rt-multi-thread"] }
tokio-stream.workspace = true
tonic-health.workspace = true
tonic-reflection.workspace = true
tonic-types.workspace = true
tonic.workspace = true
tonic-web.workspace = true
tonic-tracing-opentelemetry.workspace = true
tower.workspace = true
tower-http.workspace = true
opentelemetry.workspace = true
tracing = { workspace = true }
tracing-opentelemetry = { workspace = true }
tracing-subscriber = { workspace = true }
uuid.workspace = true
uuid7.workspace = true
base64.workspace = true
argon2 = { workspace = true }
base62.workspace = true
http = { workspace = true }
hmac-sha256 = { workspace = true }

tracing-log = { workspace = true }
secrecy.workspace = true
chacha20poly1305.workspace = true
anyhow.workspace = true
axum = { workspace = true, features = ["default", "macros", "http2"] }
hyper = { workspace = true, features = ["http1", "http2", "server"] }

rust_decimal = { workspace = true, features = ["serde_json"] }
rust_decimal_macros = { workspace = true }

metering-grpc = { workspace = true, features = ["client"] }
common-domain = { workspace = true }
common-utils = { workspace = true, features = ["error-stack-conv"] }
common-repository = { workspace = true }
distributed-lock = { workspace = true, features = ["postgres-support"] }
stripe-client = { path = "crates/stripe-client" }
object_store = { workspace = true, features = ["aws", "azure", "gcp", "http"] }
bytes.workspace = true
ring.workspace = true
async-trait.workspace = true
hex.workspace = true
url.workspace = true
tracing-futures.workspace = true
opentelemetry_sdk = { workspace = true }
fang = { workspace = true, features = ["asynk", "derive-error"] }
typetag = { workspace = true }
jsonwebtoken.workspace = true

once_cell.workspace = true
tower-layer.workspace = true
fastrand = { workspace = true }
meteroid-store.workspace = true
ndarray.workspace = true
ndarray-interp.workspace = true
rand_distr.workspace = true
rand.workspace = true
fake.workspace = true
rand_chacha.workspace = true


[dev-dependencies]
rstest = { workspace = true }
testcontainers = { workspace = true }
testcontainers-modules = { workspace = true, features = ["postgres"] }
tokio-util = { workspace = true }
clickhouse-rs = { workspace = true }
rdkafka = { workspace = true }
metering = { workspace = true }
kafka = { workspace = true }
mockito = { workspace = true }

[[bin]]
name = "meteroid-api"
path = "src/bin/server.rs"

# local / staging. For production workload, prefer running each worker as a pod and use k8s scheduler
[[bin]]
name = "meteroid-scheduler"
path = "src/bin/scheduler.rs"

[[bin]]
name = "meteroid-seeder"
path = "src/bin/seeder.rs"
