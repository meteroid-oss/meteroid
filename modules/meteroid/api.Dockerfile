
FROM lukemathwalker/cargo-chef:latest-rust-1-bookworm AS chef
ARG CARGO_REGISTRIES_CRATES_IO_PROTOCOL=sparse
ENV CARGO_NET_GIT_FETCH_WITH_CLI=true
WORKDIR /opt/src

FROM chef AS planner
COPY . .
RUN cargo chef prepare --recipe-path recipe.json

FROM chef AS builder
ARG MOLD_VERSION=1.11.0
ARG PROTO_VERSION=21.8

ARG MOLD_ARCH
ARG PROTO_ARCH
ARG GRPC_HEALTH_PROBE_ARCH

RUN test -n "$MOLD_ARCH" || (echo "MOLD_ARCH not set" && false)
RUN test -n "$PROTO_ARCH" || (echo "PROTO_ARCH not set" && false)
RUN test -n "$GRPC_HEALTH_PROBE_ARCH" || (echo "GRPC_HEALTH_PROBE_ARCH not set" && false)

RUN apt-get update

# install needed packages
RUN DEBIAN_FRONTEND=noninteractive && \
    apt-get -y install --no-install-recommends curl pkg-config unzip build-essential libssl-dev libsasl2-dev openssl cmake clang wget

# Install mold
RUN wget https://github.com/rui314/mold/releases/download/v${MOLD_VERSION}/mold-${MOLD_VERSION}-${MOLD_ARCH}-linux.tar.gz && \
    tar xvfz mold*.tar.gz && \
    mv mold*-linux/bin/* /usr/local/bin && \
    mv mold*-linux/libexec/* /usr/libexec && \
    rm -rf mold*

# Install protoc
RUN wget https://github.com/protocolbuffers/protobuf/releases/download/v${PROTO_VERSION}/protoc-${PROTO_VERSION}-linux-${PROTO_ARCH}.zip && \
    unzip protoc*.zip && \
    mv bin/protoc /usr/local/bin && \
    mv include/google /usr/local/include

# Install grpc health checker
RUN GRPC_HEALTH_PROBE_VERSION=v0.4.24 && \
    wget -qO/bin/grpc_health_probe https://github.com/grpc-ecosystem/grpc-health-probe/releases/download/${GRPC_HEALTH_PROBE_VERSION}/grpc_health_probe-linux-${GRPC_HEALTH_PROBE_ARCH} && \
    chmod +x /bin/grpc_health_probe

# Cleanup
RUN apt-get clean; \
    rm -rf /var/lib/apt/lists/*

COPY --from=planner /opt/src/recipe.json recipe.json

ARG PROFILE
ARG CI
# Build dependencies & cache
RUN cargo chef cook --recipe-path recipe.json --profile $PROFILE --package meteroid
# Build application
COPY . .
RUN cargo build -p meteroid --bin meteroid-api --profile $PROFILE


FROM debian:stable-slim
ARG PROFILE
ARG TARGET_DIR=$PROFILE
RUN apt-get update && \
    apt-get install --no-install-recommends -y ca-certificates libssl3 libsasl2-2 libpq5 && \
    rm -rf /var/lib/apt/lists/*
COPY --from=builder /opt/src/target/$TARGET_DIR/meteroid-api /usr/local/bin/meteroid-api
COPY --from=builder /bin/grpc_health_probe /bin/grpc_health_probe

RUN groupadd --system md --gid 151 \
    && useradd --system --gid md --uid 151 md

USER md

EXPOSE 50061
EXPOSE 8080

CMD ["meteroid-api"]
