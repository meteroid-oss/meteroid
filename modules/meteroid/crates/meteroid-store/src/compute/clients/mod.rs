pub mod slots;
pub mod usage;
