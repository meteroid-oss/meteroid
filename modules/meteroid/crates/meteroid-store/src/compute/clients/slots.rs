use chrono::NaiveDate;
use std::collections::HashMap;
use uuid::Uuid;

use crate::compute::errors::ComputeError;

use crate::repositories::subscriptions::SubscriptionSlotsInterface;
use common_domain::ids::TenantId;
use error_stack::{Result, ResultExt};

#[async_trait::async_trait]
pub trait SlotClient {
    async fn fetch_slots(
        &self,
        tenant_id: TenantId,
        subscription_id: &Uuid,
        component_id: &Uuid,
        // slot_unit: &String,
        invoice_date: &NaiveDate,
    ) -> Result<u32, ComputeError>;
}

#[async_trait::async_trait]
impl SlotClient for crate::Store {
    async fn fetch_slots(
        &self,
        tenant_id: TenantId,
        subscription_id: &Uuid,
        component_id: &Uuid,
        invoice_date: &NaiveDate,
    ) -> Result<u32, ComputeError> {
        let res = self
            .get_current_slots_value(
                tenant_id,
                *subscription_id,
                *component_id,
                invoice_date.clone().and_hms_opt(0, 0, 0),
            )
            .await
            .change_context(ComputeError::InternalError)?;

        Ok(res)
    }
}

pub struct MockSlotClient {
    pub data: HashMap<(Uuid, NaiveDate), u32>,
}

#[async_trait::async_trait]
impl SlotClient for MockSlotClient {
    async fn fetch_slots(
        &self,
        _tenant_id: TenantId,
        _subscription_id: &Uuid,
        component_id: &Uuid,
        invoice_date: &NaiveDate,
    ) -> Result<u32, ComputeError> {
        match self.data.get(&(*component_id, *invoice_date)) {
            Some(v) => Ok(*v),
            // None => Err(ComputeError::InternalError),
            None => Ok(0),
        }
    }
}
