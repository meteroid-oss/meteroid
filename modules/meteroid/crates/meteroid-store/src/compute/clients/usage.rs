use std::collections::HashMap;

use crate::compute::errors::ComputeError;
use crate::domain::{BillableMetric, Period};
use chrono::NaiveDate;
use common_domain::ids::{BillableMetricId, CustomerId, TenantId};
use rust_decimal::Decimal;

#[derive(Debug, Clone)]
pub struct UsageData {
    pub data: Vec<GroupedUsageData>,
    pub period: Period,
}

impl UsageData {
    pub(crate) fn single(&self) -> Result<Decimal, ComputeError> {
        if self.data.len() > 1 {
            return Err(ComputeError::TooManyResults);
        }
        Ok(self
            .data
            .first()
            .map(|usage| usage.value)
            .unwrap_or(Decimal::ZERO))
    }
}

#[derive(Debug, Clone)]
pub struct GroupedUsageData {
    pub value: Decimal,
    pub dimensions: HashMap<String, String>,
}

#[derive(Debug, Clone)]
pub struct Metadata {
    pub key: String,
    pub value: String,
}

#[async_trait::async_trait]
pub trait UsageClient: Send + Sync {
    async fn register_meter(
        &self,
        tenant_id: TenantId,
        metric: &BillableMetric,
    ) -> Result<Vec<Metadata>, ComputeError>;

    async fn fetch_usage(
        &self,
        tenant_id: TenantId,
        customer_id: CustomerId,
        customer_alias: &Option<String>,
        metric: &BillableMetric,
        period: Period,
    ) -> Result<UsageData, ComputeError>;
}

#[derive(Eq, Hash, PartialEq)]
pub struct MockUsageDataParams {
    metric_id: BillableMetricId,
    invoice_date: NaiveDate,
}

pub struct MockUsageClient {
    pub data: HashMap<MockUsageDataParams, UsageData>,
}

#[async_trait::async_trait]
impl UsageClient for MockUsageClient {
    async fn register_meter(
        &self,
        _tenant_id: TenantId,
        _metric: &BillableMetric,
    ) -> Result<Vec<Metadata>, ComputeError> {
        Ok(vec![])
    }

    async fn fetch_usage(
        &self,
        _tenant_id: TenantId,
        _customer_local_id: CustomerId,
        _customer_alias: &Option<String>,
        metric: &BillableMetric,
        period: Period,
    ) -> Result<UsageData, ComputeError> {
        let params = MockUsageDataParams {
            metric_id: metric.id,
            invoice_date: period.end,
        };
        let usage_data = self
            .data
            .get(&params)
            .cloned()
            .unwrap_or_else(|| UsageData {
                data: vec![],
                period: period.clone(),
            });
        Ok(usage_data)
    }
}

impl MockUsageClient {
    pub fn noop() -> Self {
        Self {
            data: HashMap::new(),
        }
    }
}
