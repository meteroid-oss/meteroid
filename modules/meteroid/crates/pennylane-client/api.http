### List customers/compamies https://pennylane.readme.io/v2.0/reference/getcustomers
GET https://app.pennylane.com/api/external/v2/customers HTTP/1.1
Accept: application/json
Authorization: Bearer {{api_token}}

### Create a company customer https://pennylane.readme.io/v2.0/reference/postcompanycustomer
POST https://app.pennylane.com/api/external/v2/company_customers
Content-Type: application/json
Accept: application/json
Authorization: Bearer {{api_token}}

{
  "name": "first_test_company",
  "billing_address": {
    "address": "Gates 50",
    "postal_code": "50-001",
    "city": "Wroclaw",
    "country_alpha2": "PL"
  },
  "external_reference": "cus_123456789"
}

### Update a company customer https://pennylane.readme.io/v2.0/reference/putcompanycustomer
PUT https://app.pennylane.com/api/external/v2/company_customers/157896015
Content-Type: application/json
Accept: application/json
Authorization: Bearer {{api_token}}

{
  "name": "first_test_company",
  "billing_address": {
    "address": "Gates 50",
    "postal_code": "50-001",
    "city": "Wroclaw",
    "country_alpha2": "PL"
  },
  "external_reference": "cus_123456789"
}

### customer invoices changelog https://pennylane.readme.io/v2.0/docs/tracking-data-changes-with-pennylane-api#available-changelog-endpoints
GET https://app.pennylane.com/api/external/v2/changelogs/customer_invoices HTTP/1.1
Accept: application/json
Authorization: Bearer {{api_token}}

### Upload file https://pennylane.readme.io/v2.0/reference/postfileattachments
POST https://app.pennylane.com/api/external/v2/file_attachments HTTP/1.1
Accept: application/json
Content-Type: multipart/form-data; boundary=MyBoundary
Authorization: Bearer {{api_token}}

--MyBoundary
Content-Disposition: form-data; name="file"; filename="custom_name.pdf"
Content-Type: application/pdf

< ./inv_sample.pdf
--MyBoundary--


### Import customer invoice https://pennylane.readme.io/v2.0/reference/importcustomerinvoices
POST https://app.pennylane.com/api/external/v2/customer_invoices/import HTTP/1.1
Content-Type: application/json
Accept: application/json
Authorization: Bearer {{api_token}}

{
  "file_attachment_id": 1141709,
  "external_reference": "inv_123456789",
  "invoice_number": "PL00001",
  "date": "2025-05-01",
  "deadline": "2025-05-01",
  "customer_id": 157896015,
  "currency": "EUR",
  "currency_amount_before_tax": "200.1",
  "currency_amount": "220.1",
  "amount": "220.1",
  "currency_tax": "20",
  "invoice_lines": [
    {
      "currency_amount": "220.1",
      "amount": "220.1",
      "currency_tax": "20",
      "label": "Test line",
      "quantity": 1.5,
      "raw_currency_unit_price": "20.01",
      "unit": "units",
      "vat_rate": "PL_230"
    }
  ]
}

### Mark customer invoice as paid https://pennylane.readme.io/v2.0/reference/markaspaidcustomerinvoice
PUT https://app.pennylane.com/api/external/v2/customer_invoices/1974109344/mark_as_paid
Authorization: Bearer {{api_token}}

### List customer invoices https://pennylane.readme.io/v2.0/reference/getcustomerinvoices
GET https://app.pennylane.com/api/external/v2/customer_invoices HTTP/1.1
Accept: application/json
Authorization: Bearer {{api_token}}

### Get customer by external_reference https://pennylane.readme.io/v2.0/reference/getcustomers
GET https://app.pennylane.com/api/external/v2/customers?
  filter=[{"field": "external_reference", "operator": "eq", "value": "cus_VTNnzVXRsjJuusdcEPVtl"}] HTTP/1.1
Accept: application/json
Authorization: Bearer {{api_token}}

### Get invoice by external_reference https://pennylane.readme.io/v2.0/reference/getcustomers
GET https://app.pennylane.com/api/external/v2/customer_invoices?
  filter=[{"field": "external_reference", "operator": "eq", "value": "inv_123456789"}] HTTP/1.1
Accept: application/json
Authorization: Bearer {{api_token}}

