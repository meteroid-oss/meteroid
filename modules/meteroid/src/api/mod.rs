mod cors;
pub mod server;

pub mod shared;
pub mod utils;

pub mod addons;
pub mod apitokens;
pub mod bankaccounts;
pub mod billablemetrics;
pub mod connectors;
pub mod coupons;
pub mod customers;
mod domain_mapping;
pub mod errors;
pub mod instance;
pub mod internal;
pub mod invoices;
pub mod invoicingentities;
pub mod organizations;
pub mod plans;
pub mod portal;
pub mod pricecomponents;
pub mod productfamilies;
pub mod productitems;
pub mod schedules;
pub mod sharable;
pub mod stats;
pub mod subscriptions;
pub mod taxes;
pub mod tenants;
pub mod users;
pub mod webhooksout;
