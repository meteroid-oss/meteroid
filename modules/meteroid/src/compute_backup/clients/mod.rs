pub mod subscription;
pub mod usage;
