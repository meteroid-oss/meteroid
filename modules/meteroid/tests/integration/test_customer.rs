use meteroid_grpc::meteroid::api;
use meteroid_grpc::meteroid::api::users::v1::UserRole;
use testcontainers::clients::Cli;
use tonic::Code;

use crate::helpers;
use crate::meteroid_it;
use crate::meteroid_it::container::SeedLevel;

#[tokio::test]
async fn test_customers_basic() {
    // Generic setup
    helpers::init::logging();
    let docker = Cli::default();
    let (_postgres_container, postgres_connection_string) =
        meteroid_it::container::start_postgres(&docker);
    let setup =
        meteroid_it::container::start_meteroid(postgres_connection_string, SeedLevel::MINIMAL)
            .await;

    let auth = meteroid_it::svc_auth::login(setup.channel.clone()).await;
    assert_eq!(auth.user.unwrap().role, UserRole::Admin as i32);

    let clients = meteroid_it::clients::AllClients::from_channel(
        setup.channel.clone(),
        auth.token.clone().as_str(),
        "a712afi5lzhk",
    );

    let customer_name = "friends and co".to_owned();
    let customer_alias = "fffrrriiieeennndddsss".to_owned();
    let customer_email = "fake@fake.com".to_owned();

    // create customer
    let created = clients
        .customers
        .clone()
        .create_customer(api::customers::v1::CreateCustomerRequest {
            name: customer_name.to_string(),
            alias: Some(customer_alias.to_string()),
            email: Some(customer_email.to_string()),
            billing_config: Some(api::customers::v1::CustomerBillingConfig {
                billing_config_oneof: Some(
                    api::customers::v1::customer_billing_config::BillingConfigOneof::Stripe(
                        api::customers::v1::customer_billing_config::Stripe {
                            customer_id: "customer_id".to_string(),
                            collection_method: 0,
                        },
                    ),
                ),
            }),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    let created_manual = clients
        .customers
        .clone()
        .create_customer(api::customers::v1::CreateCustomerRequest {
            name: "created_manual".to_string(),
            alias: Some("created_manual".to_string()),
            email: Some("created_manual@meteroid.com".to_string()),
            billing_config: Some(api::customers::v1::CustomerBillingConfig {
                billing_config_oneof: Some(
                    api::customers::v1::customer_billing_config::BillingConfigOneof::Manual(
                        api::customers::v1::customer_billing_config::Manual {},
                    ),
                ),
            }),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(created.name, customer_name.clone());
    assert_eq!(created.alias, Some(customer_alias.clone()));
    assert_eq!(created.email, Some(customer_email.clone()));

    // list by [fake] search
    let list_by_fake = clients
        .customers
        .clone()
        .list_customers(api::customers::v1::ListCustomerRequest {
            search: Some("fake".to_string()),
            sort_by: api::customers::v1::list_customer_request::SortBy::NameAsc as i32,
            pagination: Some(common_grpc::meteroid::common::v1::Pagination {
                limit: 10,
                offset: 0,
            }),
        })
        .await
        .unwrap()
        .into_inner()
        .customers;

    assert_eq!(list_by_fake.len(), 0);

    // list by [alias] search
    let list_by_fake = clients
        .customers
        .clone()
        .list_customers(api::customers::v1::ListCustomerRequest {
            search: Some("iiieeennn".to_string()), // part of alias
            sort_by: api::customers::v1::list_customer_request::SortBy::NameAsc as i32,
            pagination: Some(common_grpc::meteroid::common::v1::Pagination {
                limit: 10,
                offset: 0,
            }),
        })
        .await
        .unwrap()
        .into_inner()
        .customers;

    assert_eq!(list_by_fake.len(), 1);

    // get by id
    let get_by_id = clients
        .customers
        .clone()
        .get_customer_by_id(api::customers::v1::GetCustomerByIdRequest {
            id: created.id.to_string(),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(get_by_id.id, created.id.clone());
    assert_eq!(get_by_id.name, customer_name.clone());
    assert_eq!(get_by_id.alias, Some(customer_alias.clone()));

    // get by alias
    let get_by_alias = clients
        .customers
        .clone()
        .get_customer_by_alias(api::customers::v1::GetCustomerByAliasRequest {
            alias: customer_alias.to_string(),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(get_by_alias.id, created.id.clone());
    assert_eq!(get_by_alias.name, customer_name.clone());
    assert_eq!(get_by_alias.alias, Some(customer_alias.clone()));

    // patch
    let _ = clients
        .customers
        .clone()
        .patch_customer(api::customers::v1::PatchCustomerRequest {
            customer: Some(api::customers::v1::PatchCustomer {
                id: created.id.clone(),
                name: Some("new name".to_string()),
                email: None,
                alias: None,
                invoicing_email: None,
                phone: None,
                balance_value_cents: None,
                balance_currency: None,
                billing_address: None,
                shipping_address: None,
            }),
        })
        .await
        .unwrap()
        .into_inner();

    let get_by_id_patched = clients
        .customers
        .clone()
        .get_customer_by_id(api::customers::v1::GetCustomerByIdRequest {
            id: created.id.to_string(),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(get_by_id_patched.id, created.id.clone());
    assert_eq!(get_by_id_patched.name, "new name".to_string());
    assert_eq!(get_by_id_patched.alias, Some(customer_alias.clone()));

    // top up start
    let topped_up = clients
        .customers
        .clone()
        .top_up_customer_balance(api::customers::v1::TopUpCustomerBalanceRequest {
            customer_id: created.id.to_string(),
            cents: 1024,
            notes: Some("test".into()),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(topped_up.balance_value_cents, 1024);

    let topped_up = clients
        .customers
        .clone()
        .top_up_customer_balance(api::customers::v1::TopUpCustomerBalanceRequest {
            customer_id: created.id.to_string(),
            cents: -1000,
            notes: Some("test".into()),
        })
        .await
        .unwrap()
        .into_inner()
        .customer
        .unwrap();

    assert_eq!(topped_up.balance_value_cents, 24);

    let topped_up_fail = clients
        .customers
        .clone()
        .top_up_customer_balance(api::customers::v1::TopUpCustomerBalanceRequest {
            customer_id: created.id.to_string(),
            cents: -25,
            notes: Some("test".into()),
        })
        .await
        .err()
        .unwrap();

    assert_eq!(topped_up_fail.message(), "negative customer balance");
    assert_eq!(topped_up_fail.code(), Code::FailedPrecondition);

    // top up end

    // bue credits start
    let credits_invoice = clients
        .customers
        .clone()
        .buy_customer_credits(api::customers::v1::BuyCustomerCreditsRequest {
            customer_id: created_manual.id.clone(),
            cents: 1024,
            notes: Some("test".into()),
        })
        .await
        .unwrap()
        .into_inner()
        .invoice
        .unwrap();

    assert_eq!(credits_invoice.total, 1024);
    assert_eq!(credits_invoice.subtotal, 1024);
    assert_eq!(credits_invoice.applied_credits, 0);
    assert_eq!(credits_invoice.customer_id, created_manual.id);
    // bue credits end

    // teardown
    meteroid_it::container::terminate_meteroid(setup.token, setup.join_handle).await
}
