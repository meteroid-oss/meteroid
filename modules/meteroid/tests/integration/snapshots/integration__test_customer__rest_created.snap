---
source: modules/meteroid/tests/integration/test_customer.rs
expression: "&created"
---
{
  "alias": "test-customer-rest",
  "bank_account_id": null,
  "billing_address": {
    "city": "Test City",
    "country": "FR",
    "line1": "123 Test St",
    "zip_code": "12345"
  },
  "billing_email": "billing@meteroid.com",
  "currency": "EUR",
  "custom_taxes": [
    {
      "name": "Standard Tax",
      "rate": 20.0,
      "tax_code": "TAX001"
    }
  ],
  "id": "cus_xxx",
  "invoicing_emails": [
    "invoicing@meteroid.com"
  ],
  "invoicing_entity_id": "ive_xxx",
  "name": "Test Customer REST",
  "phone": "123456789",
  "shipping_address": {
    "same_as_billing": true
  },
  "vat_number": "VAT123456"
}
